db/src/db.rs

```diff
@@ -147,6 +147,14 @@ lazy_static! {
         r#"CREATE TABLE transactions (e INTEGER NOT NULL, a SMALLINT NOT NULL, v BLOB NOT NULL, tx INTEGER NOT NULL, added TINYINT NOT NULL DEFAULT 1, value_type_tag SMALLINT NOT NULL)"#,
         r#"CREATE INDEX idx_transactions_tx ON transactions (tx, added)"#,
 
+        // TODO: migration?
+        // Associate transactions to the attributes that they assert or retract.
+        r#"CREATE TABLE tx_log (tx INTEGER NOT NULL, a SMALLINT NOT NULL)"#,
```

Comment:
I don't see anywhere that you use `tx_log.a`, so you're only using this as a way to get a unique, ordered table of tx IDs.

But we already have one of those: we already have an index on `transactions.tx`. Can you remove `tx_log` here, using `transactions` instead, and save this database change for later?

Reply:
The intention is to follow-up to allow narrowing by `a`, more or less right now :)  I'm just deciding whether to take an (optional) set of attributes, which requires parsing sets, or to make the function variadic and accept an (optionally empty) set of attributes as arguments.  Usually I prefer strongly typed fixed inputs, but we don't support sets as arguments yet.

Do you have a strong preference?  (This is relevant to `(fulltext ...)` as well; there's a TODO about exactly this issue in that code.)

Reply:
As I understand it, you're strictly expressing an optimization here: that we can narrow down the set of `tx-id`s by pre-joining the `tx_log` table against a known set of attributes, rather than scanning `transactions`, right?

That rests on two assumptions, I think:

- That such filtering will reduce the space of `tx-id`s considered, often to zero. You're looking ahead to a query like

```
[:find ?vocab :where
 [(tx-data …) [[?e ?a ?v _]]]
 [(ground ?a :db.schema/vocabulary)]]
```

- That the aggregate benefit of doing a number of shorter index walks (one per joined ID) despite the cost and disk fragmentation of maintaining `tx_log`, will outweigh the cost of a longer contiguous scan of `transactions`.

Am I understanding correctly?

Reply:
Yes, but I was looking ahead to expressing that query more directly as
```
[(tx-data ... :db.schema/vocabulary) [[...]]]
```
directly.  That is, baking more of the optimization into the `tx-data` function itself, and less into the general mechanism of joining things together.

I have a hard time believing that walking the tail of transactions to find the things that are relevant to an incremental sync is a good idea.